### PR TITLE
feat: replace Logrus logger by Uber Zap

### PR DESCRIPTION
**Pull request by dunglas**
Logrus is now deprecated. It also consumes a lot of memory (#388 is maybe due to Logrus).
This PR introduces an interface making Mercure logger-agnostic, and uses Uber Zap as default implementation.

This is a first step for #366 (cc @sagikazarmark), and to migrate Mercure to a Caddy module.

### CODE REVIEW

File: hub/log.go [hunk 1]
```diff
@@ -1,44 +1,24 @@
...
 
-func createFields(u *Update, s *Subscriber) log.Fields {
-	f := addUpdateFields(log.Fields{}, u, s.Debug)
-	for k, v := range s.LogFields {
-		f[k] = v
-	}
-
-	return f
-}
+type Field = zapcore.Field
```
Comment (sagikazarmark) [inline hub/log.go:15]: This is an unfortunate consequence of Go's type system. Maybe add a comment here saying that generics will allow you to define your own contract for the `Field` interface as well.

File: hub/log.go [hunk 2]
```diff
@@ -1,44 +1,24 @@
...
-)
+import "go.uber.org/zap/zapcore"
 
-func addUpdateFields(f log.Fields, u *Update, debug bool) log.Fields {
-	f["event_id"] = u.ID
-	f["event_type"] = u.Type
-	f["event_retry"] = u.Retry
-	f["update_topics"] = u.Topics
-	f["update_private"] = u.Private
+type stringArray []string
```
Comment (sagikazarmark) [inline hub/log.go:5]: I would probably move this to a package of its own. It's an implementation and is not really related to the hub package apart from the fact that it consumes it.

**Review by sagikazarmark — COMMENTED** [review]
Just a couple ideas

File: hub/log.go [hunk 3]
```diff
@@ -1,44 +1,24 @@
...
-	case "FLUENTD":
-		log.SetFormatter(fluentd.NewFormatter())
-	}
+type Logger interface {
+	Debug(msg string, fields ...Field)
+	Info(msg string, fields ...Field)
+	Warn(msg string, fields ...Field)
+	Error(msg string, fields ...Field)
+	Fatal(msg string, fields ...Field)
+	Sync() error
```
[the remark this answers is not in the document]
Comment (dunglas) [reply hub/log.go:23]: Actually Sync is useless... except if you pass an instance of Zap which is buffered (it's not the case by default).

File: hub/log.go (hunk 2 above)
Comment (dunglas) [reply hub/log.go:5]: It's purely internal (and copied from Zap actually).

File: hub/log.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dunglas) [reply hub/log.go:23]: Actually, `Sync` isn't called in the transport constructor but when the hub is stopping to be sure to not loose any log: https://github.com/dunglas/mercure/pull/389/files#diff-898a88f8fece8d76531ea0f087e564eed09aa9719d03cdf6af6bc5ce0d7cabaaR75 (the diff generated by GitHub is misleading). This look ok to me.

[the remark this answers is not in the document]
Comment (dunglas) [reply hub/log.go:23]: Regarding `Fatal`, it is called only when the web server crashes for an unexpected reason, and we cannot do nothing about this. Fatal looks like the good level to me for this problem.

[the remark this answers is not in the document]
Comment (sagikazarmark) [reply hub/log.go:23]: I think `Sync` should be called from a `main` (or similar) function.

As for Fatal: having a separate level can be fine (although I really don't see the difference between error and fatal, both are actionable by a user), but a logger shouldn't exit an application/process. It's simply not its responsibility.

[the remark this answers is not in the document]
Comment (dunglas) [reply hub/log.go]: Ok, I remove Sync and Fatal. They are useless in our case. Thanks!

**Pull request merged**